model_compression_toolkit/core/common/hessian/trace_hessian_service.py

```diff
@@ -0,0 +1,162 @@
+# Copyright 2023 Sony Semiconductor Israel, Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+import numpy as np
+from typing import Callable, List, Any
+
+from model_compression_toolkit.core.common import Graph
+from model_compression_toolkit.core.common.hessian.trace_hessian_config import TraceHessianConfig
+from model_compression_toolkit.core.common.hessian.trace_hessian_request import TraceHessianRequest
+from model_compression_toolkit.logger import Logger
+
+
+class TraceHessianService:
+    """
+    A service to manage, store, and compute approximation of the trace of the Hessian.
+
+    This class provides functionalities to compute approximation of the trace of the Hessian matrix based on the provided configuration
+    and input images (from representative_dataset).
+    It also offers cache management capabilities for efficient computation and retrieval.
+
+    Note:
+    - The trace of the Hessian provides valuable information about the curvature of the loss function.
+    - Computation can be computationally heavy and time-consuming.
+    - The computed trace is an approximation.
+    """
+
+    def __init__(self,
+                 graph: Graph,
+                 representative_dataset: Callable,
+                 trace_hessian_configuration: TraceHessianConfig,
+                 fw_impl
+                 ):
+        """
+
+        Args:
+            graph: Float graph.
+            representative_dataset: A callable that provides a dataset for sampling.
+            trace_hessian_configuration: Configuration for trace of the Hessian approximation computation.
+            fw_impl: Framework-specific implementation for trace Hessian approximation computation.
+        """
+        self.graph = graph
+        self.representative_dataset = representative_dataset
+        self.trace_hessian_configuration = trace_hessian_configuration
+        self.fw_impl = fw_impl
+
+        self.trace_hessian_request_to_score_list = {}
+
+    def _clear_saved_hessian_info(self):
+        """Clears the saved info approximations."""
+        self.trace_hessian_request_to_score_list={}
+
+    def count_saved_info_of_request(self, hessian_request:TraceHessianRequest) -> int:
```

Comment:
Add a comment regarding what it is counting (request is computed and saved per image yada yada yada..)